dialogue-core/src/main/java/com/palantir/dialogue/core/PinUntilErrorChannel.java

```diff
@@ -0,0 +1,275 @@
+ * Upside: ensures clients will always hit warm caches
+ * Downsides:
+ * <ul>
+ *     <li>a single high volume client can result in unbalanced utilization of server nodes.
+ *     <li>when a node is restarted (and clients see 500s/connect exceptions), they will _all_ pin to a different node
+ * </ul>
+ *
+ * To alleviate the second downside, we reshuffle all nodes every 10 minutes.
+ */
+final class PinUntilErrorChannel implements LimitedChannel {
+    private static final Logger log = LoggerFactory.getLogger(PinUntilErrorChannel.class);
+
+    // we also add some jitter to ensure that there isn't a big spike of reshuffling every 10 minutes.
+    private static final Duration RESHUFFLE_EVERY = Duration.ofMinutes(10);
+
+    // TODO(dfox): could we make this endpoint-specific?
+    private final AtomicInteger currentHost = new AtomicInteger(0); // always positive
+    private final NodeList nodeList;
+
+    @VisibleForTesting
+    PinUntilErrorChannel(NodeList nodeList) {
+        this.nodeList = nodeList;
+        Preconditions.checkArgument(
+                nodeList.size() >= 2,
+                "PinUntilError is pointless if you have zero or 1 channels."
+                        + " Use an always throwing channel or just pick the only channel in the list.");
+    }
+
+    static LimitedChannel pinUntilError(List<LimitedChannel> channels) {
+        ReshufflingNodeList shufflingNodeList =
+                new ReshufflingNodeList(channels, SafeThreadLocalRandom.get(), System::nanoTime);
+        return new PinUntilErrorChannel(shufflingNodeList);
+    }
+
+    /**
+     * This strategy results in unbalanced server utilization, especially after an upstream restarts.
+     * @deprecated prefer {@link #pinUntilError}
+     */
+    @Deprecated
+    static LimitedChannel pinUntilErrorWithoutReshuffle(List<LimitedChannel> channels) {
+        ConstantNodeList constantNodeList = new ConstantNodeList(channels, SafeThreadLocalRandom.get());
+        return new PinUntilErrorChannel(constantNodeList);
+    }
+
+    @Override
+    public Optional<ListenableFuture<Response>> maybeExecute(Endpoint endpoint, Request request) {
+        int currentIndex = currentHost.get();
+        LimitedChannel channel = nodeList.get(currentIndex);
+
+        Optional<ListenableFuture<Response>> maybeFuture = channel.maybeExecute(endpoint, request);
+        if (!maybeFuture.isPresent()) {
+            OptionalInt next = incrementHostIfNecessary(currentIndex);
+            debugLogCurrentChannelRejected(currentIndex, channel, next);
+            return Optional.empty(); // if the caller retries immediately, we'll get the next host
+        }
+
+        ListenableFuture<Response> future = maybeFuture.get();
+        return Optional.of(DialogueFutures.addDirectCallback(future, new FutureCallback<Response>() {
+            @Override
+            public void onSuccess(Response response) {
+                if (response.code() >= 300) {
```

Comment:
putting this sort of logic in onSuccess handlers throughout the code is going to continue to make this framework much harder to understand than necessary. 

please consider an approach where success means _success_ always and that there should instead be a channel which coerces failures into failed futures.

why are failures here defined as >= 300, but elsewhere in code as >= 500?

Reply:
Looks like the CJR implementation only cycles when an I/O exception is thrown making a request, not based on status (aside from QoS). We wouldn't want 4xx responses to trigger rotation.